### PR TITLE
nixos/test-driver: provide a global timeout

### PR DESCRIPTION
**Pull request by RaitoBezarius**
## Description of changes

Since the debut of the test-driver, we didn't obtain a race timer with the test execution to ensure that tests doesn't run beyond a certain amount of time.

This is particularly important when you are running into hanging tests which cannot be detected by current facilities (requires more pvpanic wiring up, QMP API stuff, etc.).

Two easy examples:

- Some QEMU tests may get stuck in some situation and run for more than 24 hours → we default to 24 hours max.
- Some QEMU tests may panic in the wrong place, e.g. UEFI firmware or worse → end users can set a "reasonable" amount of time

And then, we should let the retry logic retest them until they succeed and adjust their global timeouts.

Of course, this does not help with the fact that the timeout may need to be a function of the actual busyness of the machine running the tests. This is only one step towards increased reliability.

Self tests are offered via `testBuildFailure`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
:+1: this is more robust than `meta.timeout`, e.g. when tests are accumulated with `symlinkJoin` or some reporting derivation.
Would love to use this in Nix itself to report which tests have run in which test environments. That's currently far too opaque.

> TODO: self tests?

You could write a test with a short timeout and a long sleep somewhere, then transform it with [`testers.testBuildFailure`](https://nixos.org/manual/nixpkgs/unstable/#tester-testBuildFailure), which produces a log in its `$out`.
You can then grep the log from a `runCommand` to make sure it failed in the right way (ie not due to an unrelated error).

**Comment by tfc** [comment]
General timeouts are useful and generally an improvement for CI throughput.

I have two thoughts about this change:

For how many people is a test without a timeout useful at all? In interactive mode (which could run for days for playing with), the timer is not started, which makes sense.
As much as i like "maybe" values, we could get rid of this  complexity by simply always having a timeout. And people who want no timeout can still simulate it by setting it to a week or so (and interactive users don't need to touch it at all).

The default value of 24 hours is already very relaxed. What is our knowledge about the longest legitimately running test? It might make sense to already use this CI-throughput-improving functionality by setting a useful halfway strict global timeout (as it still can be set to longer times for tests which really need it).

**Comment by RaitoBezarius** [comment]
> General timeouts are useful and generally an improvement for CI throughput.
> 
> I have two thoughts about this change:
> 
> For how many people is a test without a timeout useful at all? In interactive mode (which could run for days for playing with), the timer is not started, which makes sense.
> As much as i like "maybe" values, we could get rid of this  complexity by simply always having a timeout. And people who want no timeout can still simulate it by setting it to a week or so (and interactive users don't need to touch it at all).

Sounds reasonable!

> 
> The default value of 24 hours is already very relaxed. What is our knowledge about the longest legitimately running test? It might make sense to already use this CI-throughput-improving functionality by setting a useful halfway strict global timeout (as it still can be set to longer times for tests which really need it).

Discussed it on infra and 1 hour seems better.

File: nixos/lib/test-driver/test_driver/__init__.py [hunk 1]
```diff
@@ -76,6 +76,14 @@ def main() -> None:
         nargs="*",
         help="vlans to span by the driver",
     )
+    arg_parser.add_argument(
+        "--global-timeout",
+        type=int,
+        metavar="GLOBAL_TIMEOUT",
+        action=EnvDefault,
+        envvar="globalTimeout",
+        help="Global timeout in seconds to arm for the whole test",
```
Comment (nikstur) [inline nixos/lib/test-driver/test_driver/__init__.py:85]:
```suggestion
        help="Timeout in seconds for the whole test",
```

File: nixos/lib/testing/driver.nix [hunk 2]
```diff
@@ -123,6 +124,18 @@ in
       defaultText = "hostPkgs.qemu_test";
     };
 
+    globalTimeout = mkOption {
+      description = mdDoc ''
+        A global timeout for the complete test, expressed in seconds.
+        Beyond that timeout, every resource will be killed and released and the test will fail.
+
+        By default, we use a 24 hours timeout.
```
Comment (nikstur) [inline nixos/lib/testing/driver.nix:132]: 24 sounds very long. What about 6 hours? Do we have any tests that run longer? If they do it's probably better if they manually override the timeout anyway.

**Review by nikstur — COMMENTED** [review]
Very very cool

**Review by blitz — APPROVED** [review]
I love it! :heart:

**Comment by RaitoBezarius** [comment]
For self testing, I am reviving #157161 and plan to use it as a base for the self test here.

**Comment by RaitoBezarius** [comment]
@roberth Done for everything in this scope. I would suggest to get this first probably. :)

File: nixos/lib/test-driver/test_driver/driver.py [hunk 3]
```diff
@@ -161,6 +172,12 @@ def join_all(self) -> None:
         with rootlog.nested("wait for all VMs to finish"):
             for machine in self.machines:
                 machine.wait_for_shutdown()
+            self.race_timer.cancel()
+
+    def terminate_test(self) -> None:
+        with rootlog.nested("timeout reached; test termination..."):
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/driver.py:178]:
```suggestion
        with rootlog.nested("timeout reached; test terminating..."):
```

File: nixos/tests/nixos-test-driver/timeout.nix [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+  name = "Test that sleep of 6 seconds fails a timeout of 5 seconds";
+  globalTimeout = 5;
+
+  nodes = {
+    machine = ({ pkgs, ... }: {
+      environment.systemPackages = [
+        pkgs.busybox
+      ];
```
Comment (roberth) [inline nixos/tests/nixos-test-driver/timeout.nix:9]:
```suggestion
```

Don't we have `sleep` from `coreutils`?

File: nixos/lib/test-driver/test_driver/driver.py [hunk 5]
```diff
@@ -144,6 +151,10 @@ def test_script(self) -> None:
 
     def run_tests(self) -> None:
         """Run the test script (for non-interactive test runs)"""
+        rootlog.info(
+            f"Test will timeout and terminate in {self.global_timeout} seconds"
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/driver.py:155]:
```suggestion
            f"Test will time out and terminate in {self.global_timeout} seconds"
```

verb

File: nixos/lib/test-driver/test_driver/driver.py [hunk 6]
```diff
@@ -161,6 +172,12 @@ def join_all(self) -> None:
         with rootlog.nested("wait for all VMs to finish"):
             for machine in self.machines:
                 machine.wait_for_shutdown()
+            self.race_timer.cancel()
+
+    def terminate_test(self) -> None:
+        with rootlog.nested("timeout reached; test termination..."):
+            for machine in self.machines:
+                machine.release()
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/driver.py:180]: This doesn't seem to handle non-termination in the `testScript`.
Also what happens if the `testScript` has a `try`/`except`? The option doc claims that the script will fail, but this does not seem to be guaranteed.

Comment (RaitoBezarius) [reply nixos/lib/test-driver/test_driver/driver.py:180]: What do you mean exactly by non-termination? The test passes? Termination is done somewhere else (by releasing all the machines too with a different message).

Good point I guess about the `try/except`, I wonder if there's any form of non-catchable exception inside Python but everyone would break down in general.

In any case, I guess we could ensure termination by exiting with exit code 2 (to make it clear it's a timeout-driven exit).

Comment (RaitoBezarius) [reply nixos/lib/test-driver/test_driver/driver.py:183]: Here's a proposal, I will kill the main thread in the termination sequence.

Comment (RaitoBezarius) [reply nixos/lib/test-driver/test_driver/driver.py:183]: (more precisely: I will kill the process and let the kernel do the right thing.)

File: nixos/lib/testing/run.nix [hunk 7]
```diff
@@ -29,25 +38,26 @@ in
...
-    test = lib.lazyDerivation { # lazyDerivation improves performance when only passthru items and/or meta are used.
-      derivation = hostPkgs.stdenv.mkDerivation {
-        name = "vm-test-run-${config.name}";
+    rawTestDerivation = hostPkgs.stdenv.mkDerivation {
+      name = "vm-test-run-${config.name}";
 
-        requiredSystemFeatures = [ "kvm" "nixos-test" ];
+      requiredSystemFeatures = [ "kvm" "nixos-test" ];
 
-        buildCommand = ''
```
Comment (RaitoBezarius) [inline nixos/lib/testing/run.nix:38]: Ugh, I should maybe get rid of those unrelated changes.

File: nixos/lib/test-driver/test_driver/driver.py (hunk 6 above)
Comment (roberth) [reply nixos/lib/test-driver/test_driver/driver.py:183]: > kill the process

sgtm.

Alternatively, python may have a direct exit "syscall" that doesn't throw an exception but just does the thing immediately.

Comment (roberth) [reply nixos/lib/test-driver/test_driver/driver.py:183]: This seems to be it: [os._exit](https://docs.python.org/3/library/os.html#os._exit)

Comment (RaitoBezarius) [reply nixos/lib/test-driver/test_driver/driver.py:183]: Oh right, I guess using `kill` is less obscure than `_exit`, if that's OK with you.

Comment (roberth) [reply nixos/lib/test-driver/test_driver/driver.py:183]: Not sure if that's so great though. `sys.exit` might throw an exception and therefore not be airtight, but you don't get any cleanup either. Script author won't be able to clean up out-of-process resources, such as child processes.
Maybe it should just be `sys.exit` anyway, although I don't know how that exception would propagate to the main thread?

Comment (roberth) [reply nixos/lib/test-driver/test_driver/driver.py:183]: Usually we have Nix to fall back on as a kind of process manager, but that's not the case with interactive use (`nix run .#foo.driverInteractive`).

Comment (RaitoBezarius) [reply nixos/lib/test-driver/test_driver/driver.py:183]: For non-interactive usecase, this is exactly what we want, I would say.
For interactive usecase, well… it was ruled out of scope from the get-go and I would argue that timeout should be disabled in the interactive usecase.

Using `sys.exit` is impossible because there's no way we can propagate the `threading.Timer` exception back to the main thread.

We would need our own way to control error handling and offer a chance to terminate gracefully, I would expect this would require some serious back'n'forth in the code to achieve this, including running the test script on its own thread and having the main thread as a manager.

AFAIK, out of process resources should still be killed in interactive because they are spawned off from a unique hierarchy (the interactive runner).

So… I would argue this is as good as we can go without investing much more time, which I am afraid I won't be able to commit to this in particular.

Comment (roberth) [reply nixos/lib/test-driver/test_driver/driver.py:183]: > there's no way we can propagate

Makes sense actually. Usually only languages that exercise tight control over their runtime allow this, examples being Java and Haskell.
(Also pthreads I guess, but unsafely)

Sounds good enough to me then.

**Review by roberth — APPROVED** [review]

File: nixos/lib/testing/run.nix (hunk 7 above)
Comment (RaitoBezarius) [reply nixos/lib/testing/run.nix:38]: Ah they are needed…
